tests/subsys/bluetooth/fast_pair/storage/factory_reset/src/main.c

```diff
@@ -0,0 +1,516 @@
+#include <zephyr/settings/settings.h>
+#include <zephyr/sys/reboot.h>
+#include <zephyr/storage/flash_map.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(test_main);
+
+#include <bluetooth/services/fast_pair.h>
+
+#include "fp_storage_ak.h"
+#include "fp_storage_ak_priv.h"
+#include "fp_storage_pn.h"
+#include "fp_storage_pn_priv.h"
+#include "fp_storage_manager.h"
+#include "fp_storage_manager_priv.h"
+
+#include "common_utils.h"
+
+
+#define ACCOUNT_KEY_MAX_CNT CONFIG_BT_FAST_PAIR_STORAGE_ACCOUNT_KEY_MAX
+
+#define SETTINGS_RESET_TEST_SUBTREE_NAME "fp_reset_test"
+#define SETTINGS_STAGE_KEY_NAME "stage"
+#define SETTINGS_NAME_SEPARATOR_STR "/"
+#define SETTINGS_RESET_TEST_STAGE_FULL_NAME \
+	(SETTINGS_RESET_TEST_SUBTREE_NAME SETTINGS_NAME_SEPARATOR_STR SETTINGS_STAGE_KEY_NAME)
+
+BUILD_ASSERT(SETTINGS_NAME_SEPARATOR == '/');
+
+enum reset_test_stage {
+	RESET_TEST_STAGE_NORMAL,
+	RESET_TEST_STAGE_INTERRUPTED,
+};
+
+enum reset_test_finished {
+	RESET_TEST_FINISHED_SECTION_PLACEMENT_BIT_POS,
+	RESET_TEST_FINISHED_STORE_BIT_POS,
+	RESET_TEST_FINISHED_NORMAL_RESET_BIT_POS,
+	RESET_TEST_FINISHED_INTERRUPTED_RESET_BEGIN_BIT_POS,
+
+	RESET_TEST_FINISHED_COUNT,
+};
+
+static enum reset_test_stage test_stage = RESET_TEST_STAGE_NORMAL;
+static uint8_t test_finished_mask;
+
+BUILD_ASSERT(__CHAR_BIT__ * sizeof(test_finished_mask) >= RESET_TEST_FINISHED_COUNT);
+
+static int settings_set_err;
+
+static bool storage_reset_pass;
+
+
+static int settings_load_stage(size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	int rc;
+
+	if (len > sizeof(test_stage)) {
+		return -EINVAL;
+	}
+
+	rc = read_cb(cb_arg, &test_stage, len);
+	if (rc < 0) {
+		return rc;
+	}
+
+	if (rc != len) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int fp_reset_test_settings_set(const char *name, size_t len, settings_read_cb read_cb,
+				      void *cb_arg)
+{
+	int err = 0;
+
+	if (!strncmp(name, SETTINGS_STAGE_KEY_NAME, sizeof(SETTINGS_STAGE_KEY_NAME))) {
+		err = settings_load_stage(len, read_cb, cb_arg);
+	} else {
+		err = -ENOENT;
+	}
+
+	/* The first reported settings set error will be remembered by the module.
+	 * The error will then be propagated by the commit callback.
+	 * Errors returned in the settings set callback are not propagated further.
+	 */
+	if (err && !settings_set_err) {
+		settings_set_err = err;
+	}
+
+	return err;
+}
+
+static int fp_reset_test_settings_commit(void)
+{
+	if (settings_set_err) {
+		return settings_set_err;
+	}
+
+	return 0;
+}
+
+SETTINGS_STATIC_HANDLER_DEFINE(fp_reset_test, SETTINGS_RESET_TEST_SUBTREE_NAME, NULL,
+			       fp_reset_test_settings_set, fp_reset_test_settings_commit, NULL);
+
+static void mark_test_as_finished(enum reset_test_finished bit_pos)
+{
+	WRITE_BIT(test_finished_mask, bit_pos, 1);
+}
+
+static int test_fp_storage_reset_perform(void)
+{
+	if (storage_reset_pass) {
+		mark_test_as_finished(RESET_TEST_FINISHED_INTERRUPTED_RESET_BEGIN_BIT_POS);
+		storage_reset_pass = false;
+		ztest_test_pass();
```

Comment:
Nice trick! What's stopping us from removing the sys_reboot call altogether and adding the ```test_interrupted_reset_finish``` to the first test suite? I guess we can simulate the reboot behaviour in the setup function of the ```test_interrupted_reset_finish``` test case by clearing the RAM of the affected modules

Reply:
We don't know what is the state of settings subsystem internal variables/mutexes etc. We can not simulate reset in settings module.

Reply:
We are not interrupting the execution during the ```settings_load``` but during the ```bt_fast_pair_factory_reset```